### PR TITLE
Add retry for data store properties access; new property to enable use of fallbacks

### LINKED ISSUE #2075: `DataStorePropertiesConfiguration` needs to do retries, fail only after multiple retries
Currently Stargate V2 services will access persistence (backend) metadata during startup; this is done by class `DataStorePropertiesConfiguration`. It will only try one call to Bridge, and if that fails it gives up (logging stack trace) but continues with some arbitrary defaults which are unlikely to be correct values for the actual backend.

Instead, what should happen is that this class attempted call multiple times with some delay in-between, and if exhausting maximum time/attempts, fail to start up.

### PR DESCRIPTION
**Pull request by tatu-at-datastax**
**What this PR does**:

Improves reliability of data store metadata lookup on startup; makes use of possible fallbacks configurable (defaulting to no fallbacks)

**Which issue(s) this PR fixes**:
Fixes #2075

**Checklist**
- [x] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/config/DataStoreConfig.java [hunk 1]
```diff
@@ -33,6 +33,13 @@ public interface DataStoreConfig {
   @WithDefault("${stargate.multi-tenancy.enabled}")
   boolean ignoreBridge();
 
+  /**
+   * @return If call(s) to fetch metadata fail, should we just return default settings ({@code
+   *     true}) or throw an exception ({@code false}). Defaults to {@code false}.
+   */
+  @WithDefault("false")
+  boolean useFallbacksIfCallFails();
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/config/DataStoreConfig.java:41]: maybe something like `bridgeFallbackEnabled()`, the other one is really hard to get with `-D` or evn vars

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java [hunk 2]
```diff
@@ -48,25 +50,51 @@ DataStoreProperties configuration(
...
+      delay = 5,
+      delayUnit = ChronoUnit.SECONDS,
+      maxDuration = 60,
+      durationUnit = ChronoUnit.SECONDS)
+  private Schema.SupportedFeaturesResponse fetchSupportedFeatures(
+      StargateBridgeGrpc.StargateBridgeBlockingStub bridge) {
+    try {
+      return bridge.getSupportedFeatures(Schema.SupportedFeaturesRequest.newBuilder().build());
+    } catch (Exception e) {
+      LOG.warn("Data store metadata fetch using Bridge failed, problem: {}", e.getMessage());
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:96]: maybe on info level, and warn only if it completely fails

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java [hunk 3]
```diff
@@ -48,25 +50,51 @@ DataStoreProperties configuration(
...
+   * Method for fetching data store metadata: will attempt up to 5 retries, for up to 1 minute
+   * before failing.
+   */
+  @Retry(
+      maxRetries = 5,
+      delay = 5,
+      delayUnit = ChronoUnit.SECONDS,
+      maxDuration = 60,
+      durationUnit = ChronoUnit.SECONDS)
+  private Schema.SupportedFeaturesResponse fetchSupportedFeatures(
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:91]: we should extend the `DataStorePropertiesConfiguration` and add a case where first call fails and second one works..

**Review by ivansenic — APPROVED** [review]

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/config/DataStoreConfig.java (hunk 1 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/config/DataStoreConfig.java:41]: Yeah, good point. Will change as suggested.

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java (hunk 2 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:96]: Hmmh. I can see it going either way, but sure I can change this to `info`.

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java (hunk 3 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java]: @ivansenic Yes; was able to do the test based on existing mock and some googling on how to override `@Retry` defaults. It is impressive how various things can be overridden.

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java [hunk 4]
```diff
@@ -48,25 +51,65 @@ DataStoreProperties configuration(
...
+      throw e;
+    }
+  }
+
+  /**
+   * Method for fetching data store metadata: will attempt up to 9 retries, for up to 1 minute
+   * before failing.
+   */
+  @Retry(
+      maxRetries = 9,
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:93]: is `9` really needed? I mean even `5` seems too high to me..

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java [hunk 5]
```diff
@@ -48,25 +51,65 @@ DataStoreProperties configuration(
...
+      delay = 3,
+      delayUnit = ChronoUnit.SECONDS,
+      maxDuration = 60,
+      durationUnit = ChronoUnit.SECONDS)
+  // NOTE: Must NOT be 'private' for annotation to take effect!
+  protected Schema.SupportedFeaturesResponse fetchSupportedFeatures(
+      StargateBridgeGrpc.StargateBridgeBlockingStub bridge, AtomicInteger callCount) {
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:100]: why the `AtomicInteger` here? log only purpose? I guess this is over engineered :-P

File: apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * Tests for verifying retry logic of {@link DataStorePropertiesConfiguration#configuration} call.
+ *
+ * <p>NOTE: properties of {@code @Retry} annotation are overridden to both reduce delay between
+ * calls AND to lower max-retry to 2 calls. See "./src/test/resources/application.yaml" for details.
+ * Override logic by Quarkus/SmallRye is explained <a
+ * href="https://quarkus.io/guides/smallrye-fault-tolerance">here</a>.
+ */
+@QuarkusTest
+public class BridgeDataStorePropertiesRetryTest extends BridgeTest {
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java:35]: nit: Oh man understanding this test was so hard for me.. This is why I try not to have de-duplication in the test code.. It's almost impossible to read the tests and go through the code with understanding this easily.. Just imagine how this works:

1. Start at the test case -> `dataStoreWithNoRetriesOk() `
2. OK, just delegates to the method, `dataStoreWithNCalls()`, ok that's at the bottom of the class, navigate, understand parameters, etc
3. And now this calls another method `mockGetSupportedFeaturesCall()` which is now top of the class :exploding_head:

Clear `Given -> When -> Then` structure is way better.. Can you at least kill the `mockGetSupportedFeaturesCall` and push all code to the `dataStoreWithNCalls`, there's really no reason to split this up..

File: apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  @Test
+  public void dataStoreWithTwoRetriesFail() {
+    // Fails if 3 retries needed (test setup only allows 2)
+    try {
+      dataStoreWithNCalls(4);
+      Assertions.fail("Should not have succeeded (max 2 retries)");
+    } catch (Exception e) {
+      assertThat(e).isInstanceOf(StatusRuntimeException.class);
+      assertThat(e.getMessage()).contains("UNAVAILABLE");
+    }
```
Comment (ivansenic) [inline apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java:94]:
```
Throwable e = catchThrowable(() ->  dataStoreWithNCalls(4));

assertThat(e).isInstanceOf(StatusRuntimeException.class);
assertThat(e.getMessage()).contains("UNAVAILABLE");
```

**Review by ivansenic — APPROVED** [review]

**Comment by ivansenic** [comment]
Formatting fixes required as well..

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java (hunk 4 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:93]: Round number like 10,  but mostly that it'd try for closer to 1 minute, but without super long delay in-between.
We can change it of course. Depends on what kind of problem/outage we'd have.

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java (hunk 5 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:100]: Yes, just for logging. I like seeing how many retries were made, and which call is which.
I don't think the goal is to absolutely minimize code.

File: apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java (hunk 6 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java:35]: Ok, I am sorry I caused you headache :)
I did go through couple of rounds of refactoring, limited by this being bit new code to me (wrt mocking). I am bit surprised that it is so hard to follow, given that amount of code is not super high, but YMMV. But I am sure I can improve ordering, structure based on comments above.

File: apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java (hunk 4 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/main/java/io/stargate/sgv2/api/common/properties/datastore/configuration/DataStorePropertiesConfiguration.java:93]: I'll change this back to 5 total calls (4 retries); can tune later on if need be.

File: apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java (hunk 6 above)
Comment (tatu-at-datastax) [reply apis/sgv2-quarkus-common/src/test/java/io/stargate/sgv2/api/common/properties/datastore/configuration/BridgeDataStorePropertiesRetryTest.java:33]: @ivansenic Ok here goes again. I hope it's bit more readable now.

**Comment by ivansenic** [comment]
@tatu-at-datastax I already approved this, you can ship it.. No need to wait anymore on my side..

**Comment by tatu-at-datastax** [comment]
@ivansenic yes, I saw approval but since this isn't urgent wanted to give you a chance just in case you wanted to add some more suggestions. Thanks!
